postgresql: more cleanup

wolfgangwalther opened the pull request:
## Description of changes

TLDR: This is a continuation of #292993 including those changes which will cause rebuilding the postgresql derivations. Only cleanup, no functional changes.

While working on building PostgreSQL with meson as mentioned in #292993, I had to carefully look at every line in the old builder to see if and how to take this to the meson build. Some of the special cases seem to be not needed anymore, so this PR removes them. Trying to understand the purpose of each of the custom patches, I also cleaned those up along the way with some common formatting and improved naming. Hopefully, this would not only help my own understanding, but other's, too. Finally this PR also contains one more change that is related to the refactor of `this` in the other PR - here, I remove `thisAttr`, too, and make sure that `.tests` will always work correctly, even when overriding the derivation.

~~This merges into the branch of #292993, so please ignore the first 13 or so commits (all starting with "refactor ...") - those are discussed in the other PR.~~ I still split this into a second PR, because the other changes are a pure refactor without causing rebuilds - while those changes there will cause a full rebuild of all postgresql derivations. Thus, I also targeted staging instead of master with this PR here.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Maintainers: @thoughtpolice @danbst @globin @marsam @ivan @Ma27

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
postgresql: clean up patches

More cosmetic than anything else. The description in findstring.patch was
duplicated, postgresql-9.6.1 references confusing. Timestamps are not
needed, that's what git blame is for.

wolfgangwalther committed:
postgresql: remove thisAttr argument by calling tests directly

Previously, it was not possible to run tests on an overridden derivation, because
the derivation under test was always pulled from pkgs.

With this change, the following will return the same test:

  postgresql_jit.tests

and

  (postgresql.override { jitSupport = true; }).tests

wolfgangwalther committed:
postgresql: change pname for postgresql_jit attributes to postgresql-jit

This is for consistency between the top-level attribute and the respective package
name, which makes it easier to track derivations.

wolfgangwalther committed:
postgresql: rename patches

Trying to understand what each patch does made me come up with some more
descriptive names:

- Renaming the two disable-xxx patches to a common names makes it immediately
  clear that one replaces the other depending on version number.
- findstring was really not descriptive at all.
- hardcode-pgxs-path will be extended with more paths for split outputs in
  a later commit. Renaming here already to allow git to better track renames.

Finally replacing HARDCODED_PGXS_PATH with $out/lib in the last patch, makes
it easier to understand what the end result will look like when reading the
patch.

wolfgangwalther committed:
postgresql: remove left-over LC_ALL environment variable

This was introduced in 65fd8f3f48fed5e2a4b82da7e354fac53ae72099 around 20 years
ago as LANG=en_US, later changed to LC_ALL=en_US and LC_ALL=C to avoid bash
warnings. This might have been relevant to run initdb during the check phase, but
the PostgreSQL docs [1] say this today:

  For implementation reasons, setting LC_ALL does not work for this purpose

Therefore it can be removed safely.

[1]: https://www.postgresql.org/docs/current/regress-run.html#REGRESS-RUN-LOCALE

wolfgangwalther committed:
postgresql: move explicit libxml2 include behind version flag

This was introduced in #44083 to fix cross building, where xml2-config
wouldn't run on the host platform. This was fixed upstream two years
later [1], so that from v13 on pkg-config is used before xml2-config is.

Once v12 is EOL, we can remove this entirely.

[1]: https://github.com/postgres/postgres/commit/0bc8cebdb889368abdf224aeac8bc197fe4c9ae6

wolfgangwalther committed:
postgresql: enable parallel building on darwin

This was disabled to fix #51093. The related discussion upstream revealed that this
has been fixed in [1] for v12+. Since v12 is our oldest supported version right now,
we should be able to safely enable parallel building on darwin again.

Furthermore, it seems that this kind of issue was also fixed in #51221 and #51408
for the general case.

[1]: https://github.com/postgres/postgres/commit/826eff57c4c23f77314ba7151d3dc506ce0fa24c

wolfgangwalther committed:
postgresql: remove preConfigure CC variable

This seems to have been done better by now. There are no references kept
and the build does not fail when removing it.

wolfgangwalther committed:
pkgsMusl.postgresql: mark v12 and v13 with JIT as broken

Those currently don't build on master, not sure whether they ever built
correctly in the first place.

Ma27 commented:
For now, just one request: please don't @ me in commit messages (feel free to reference my GH handle or my name, but without @). The reason is that too many people out there have interesting rebase workflows in their forks that cause notifications in my inbox whenever they push something (had this issue a few times already) and it was always pretty annoying.

wolfgangwalther committed:
git-blame-ignore-revs: add postgresql cleanup

Ma27 commented:
Just to make sure it doesn't get lost, from https://github.com/NixOS/nixpkgs/pull/292993#issuecomment-1987201845:

would you mind trying out if adding 2da789a to .git-blame-ignore-revs helps to retain a bit of the history of the previous file?

No idea if that works out, but maybe worth a try?

wolfgangwalther committed:
nixos/postgresql: fix enableJIT = false

The enableJIT = true case was fixed in #221851 or
e2fb65175228a992f196f3b1700a53e18602e7f6 respectively.

However this did not take the case into consideration, when doing this:

    services.postgresql = {
      enable = true;
      enableJIT = false;
      package = pkgs.postgresql_15_jit;
    };

If enableJIT is treated as the source of truth, then this should indeed
cause JIT to be disabled, which this commit does.

wolfgangwalther committed:
pkgsMusl.postgresql: add TODO to run all regression tests

This was fixed upstream as a result of this discussion:

https://www.postgresql.org/message-id/flat/ef1b47df-0808-4bd7-b08c-5153d5d75f4c%40technowledgy.de

wolfgangwalther committed:
pkgsMusl.postgresql: remove icu-collations-hack patch

This was introduced in #228349, but doesn't seem necessary to build for pkgsMusl.

In fact the patch itself seems to be very specific about a stripped down icu
package in Alpine Linux, which does not apply to nixpkgs.

Ma27 commented:
I think this needs another rebase.

marsam commented on nixos/tests/postgresql-wal-receiver.nix:
```diff
@@ -9,111 +10,117 @@ let
+
+      recoveryFile = if atLeast12
+          then pkgs.writeTextDir "recovery.signal" ""
+          else pkgs.writeTextDir "recovery.conf" "restore_command = 'cp ${walBackupDir}/%f %p'";
```
nitpick, we no longer have PostgreSQL<12 in nixpkgs, so I think we can remove this branch

wolfgangwalther committed:
nixosTests.postgresql-wal-receiver: remove left-over v12 conditionals

We don't have postgresql < 12 in nixpkgs anymore.

Ma27 commented:
> Can I do anything to help the review process?

Not on my end unfortunately, it's mostly an issue of time vs. other stuff I care about. That said, this PR and another larger change I'd really like to see moved forward are pretty high up on my todo list, so I'd like to assure you that I haven't forgotten about this and that I'll do my best to help bringing this forward soonish.

marsam reviewed (approved):
LGTM. Great work, thanks a lot!

Ma27 commented:
Looking good. Would like to run a few tests on my own with it before approving and merging.

Btw cc @yu-re-ka would you mind taking a look at the musl-related patches?

Nice work!

Ma27 commented:
Apologies for the unnecessary ping then!

I thought there was discussion around that, but I didn't find it in this thread, sorry.
